### ISSUE TITLE
Build succeeds with 0.3.1 but fails with 0.4.0

### ISSUE THREAD

**Issue by wsoula** [gomods/athens#1219]
**Describe the bug**
I have a developer that opened a pull request after I upgraded athens to 0.4.0 and his pr failed even though he is not changing dependencies and just go code.  I rolled back athens to 0.3.1 and it succeeded, I then upgraded to 0.4.0 again and it failed, I then rolled back to 0.3.1 and it succeeded.  If I click the link in the error message it returns: "{"Version":"v1.0.0-comfg","Time":"2018-05-31T13:18:12Z"}"

**Error Message**
```console
go: finding go.opencensus.io v0.18.1-0.20181204023538-aab39bd6a98b
go: git.drillinginfo.com/DI2Schemas/geo@v1.0.0-comfg: unexpected status (https://athens.dev.drillinginfo.com/git.drillinginfo.com/%21d%21i2%21schemas/geo/@v/v1.0.0-comfg.info): 404 Not Found
go: finding google.golang.org/genproto v0.0.0-20180831171423-11092d34479b
```
</pre>

**To Reproduce**
Steps to reproduce the behavior: it looks like the build is running go test and looking at the output it is finding the dependencies

**Expected behavior**
For it to find the dependency

**Environment (please complete the following information):**
 - OS: linux
 - Go version : 2.1.0
 - Proxy version : 0.4.0
 - Storage (fs/mongodb/s3 etc.) : fs

**Additional context**
I'm not certain how to get additional information about the error.  Let me know what I can do to help get more information needed to debug this issue.

**Comment by marwan-at-work** [comment]
@wsoula do you have a `go.mod` that I can use to reproduce this error? Thanks!

**Comment by marwan-at-work** [comment]
I just built Athens with an `fs` disk in version v0.3.1, then rebuilt it with the same storage (populated by v0.3.1) but with v0.4.0 and it built ok.

**Comment by wsoula** [comment]
I looked into this more and deleted some files from the geo folder and it works now.  I'm sorry I was wrong about the storage its s3.

With these files in the geo folder the build works fine:
![image](https://user-images.githubusercontent.com/1820481/57933855-8ccbfc80-787b-11e9-9863-8e80f580e7b8.png)

But if either of these sets are in that folder it does not work:
![image](https://user-images.githubusercontent.com/1820481/57933899-a79e7100-787b-11e9-9a8e-ff338f226090.png)

I figured this out by moving the geo folder to the root of the bucket and letting the build pull in what it needed, then copying in sets of files till the build failed, then removing that set and copying the next and seeing if that failed.  the v0.0.0 is fine, but the other v1.0.0 causes it to fail.  I ran the build against v0.3.1 of athens first to make sure the failure still occurred.  

I'm happy to close this since I'm not blocked and can use athens v0.4.0 now.  Also, I can post the go.mod from the repo if you want, but I don't how much good it would do since it uses a bunch of our private repos.

**Comment by wsoula** [comment]
How do those "intermediate" files get created?  They keep reappearing and I have to delete them to fix the build again.

**Comment by marwan-at-work** [comment]
@wsoula there shouldn't be any intermediate files. Athens only saves 3 files per module (.info, .mod, and .zip)

**Comment by wsoula** [comment]
The long numbered v1.0.0 are coming from someone's dependencies then I guess?

**Comment by marwan-at-work** [comment]
@wsoula most likely as long as they are valid tags/metatags.

**Comment by wsoula** [comment]
I have steps to reproduce with public repos:
Delete local cache and long numbered v1.0.0 from s3 bucket
go.mod:
```
module github.com/wsoula/test

require github.com/wsoula/jenkins-job-builder v1.0.0-comfg.0.20141216225024-1d3af0f3ceaf

//require github.com/wsoula/jenkins-job-builder v1.0.0-comfg
```
then i run:
```
go get
go: finding github.com/wsoula/jenkins-job-builder v1.0.0-comfg.0.20141216225024-1d3af0f3ceaf
```
then i change go.mod to:
```
module github.com/wsoula/test

//require github.com/wsoula/jenkins-job-builder v1.0.0-comfg.0.20141216225024-1d3af0f3ceaf

require github.com/wsoula/jenkins-job-builder v1.0.0-comfg
```
then i run and get the error:
```
go get
go: finding github.com/wsoula/jenkins-job-builder v1.0.0-comfg
go: github.com/wsoula/jenkins-job-builder@v1.0.0-comfg: unexpected status (https://athens.dev.drillinginfo.com/github.com/wsoula/jenkins-job-builder/@v/v1.0.0-comfg.info): 404 Not Found
go: error loading module requirements
```
Then I delete the long numbered files from the s3 bucket and run:
```
go get
go: finding github.com/wsoula/jenkins-job-builder v1.0.0-comfg
```

**Comment by marwan-at-work** [comment]
@wsoula Thanks for giving us a way to reproduce it. That makes a lot of sense now. The reason this is failing is because we check modules in v0.4.0 based on prefixes not based on explicit paths, so when two Versions have the same prefix, we get into trouble: 

1. Here's the code now: https://github.com/gomods/athens/blob/master/pkg/storage/s3/checker.go#L26
2. Here's the code in v0.3.0: https://github.com/gomods/athens/blob/v0.3.1/pkg/storage/s3/checker.go#L28

We should probably fix this and release a v0.4.1 patch fix. 

CC: @marpio
